Phi3 runner uses TextLLMRunner

larryliu0820 opened the pull request:
As titled.

mergennachin commented on extension/tensor/tensor_ptr_maker.h:
```diff
@@ -683,5 +683,41 @@ inline TensorPtr randint(
   return randint_strided(low, high, std::move(sizes), {}, type, dynamism);
 }
 
+/**
+ * Creates a 2-D tensor (sizes=[1, max]) with values from `start` to `end`
+ * (exclusive) with step size `step`.
+ *
+ * @param start The starting value of the sequence.
+ * @param end The ending value of the sequence (exclusive).
+ * @param step The step size between values in the sequence.
+ * @param type The scalar type of the tensor elements.
+ * @param dynamism Specifies whether the tensor's shape is static or dynamic.
+ * @return A TensorPtr instance managing the newly created Tensor.
+ */
+TensorPtr arange(
+    executorch::aten::Scalar start,
+    executorch::aten::Scalar end,
+    executorch::aten::Scalar step = 1,
+    executorch::aten::ScalarType type = executorch::aten::ScalarType::Float,
+    executorch::aten::TensorShapeDynamism dynamism =
+        executorch::aten::TensorShapeDynamism::DYNAMIC_BOUND);
+
+/**
+ * Creates a 2-D tensor (sizes=[1, max]) with values from 0 to `end` (exclusive)
+ * with step size 1.
+ *
+ * @param end The ending value of the sequence (exclusive).
+ * @param type The scalar type of the tensor elements.
+ * @param dynamism Specifies whether the tensor's shape is static or dynamic.
+ * @return A TensorPtr instance managing the newly created Tensor.
+ */
+inline TensorPtr arange(
+    executorch::aten::Scalar end,
+    executorch::aten::ScalarType type = executorch::aten::ScalarType::Float,
+    executorch::aten::TensorShapeDynamism dynamism =
+        executorch::aten::TensorShapeDynamism::DYNAMIC_BOUND) {
+  return arange(0, end, 1, type, dynamism);
+}
+
```
@shoumikhin can you double check the API and the implementaiton?

mergennachin commented on extension/llm/runner/text_prefiller.h:
```diff
@@ -21,7 +21,8 @@ class ET_EXPERIMENTAL TextPrefiller {
       TextDecoderRunner* text_decoder_runner,
-      bool use_kv_cache_,
+      bool use_kv_cache,
+      bool use_cache_positions,
```
Create docblock. I presume TextPrefiller is a public API, right?

mergennachin commented on extension/llm/runner/text_llm_runner.cpp:
```diff
@@ -300,7 +302,24 @@ std::unordered_map<std::string, int64_t> get_llm_metadata(
+  // Check if we are using cache positions instead of input pos.
+  auto second_input_info = method_meta.input_tensor_meta(1).get();
+  // For input_pos, size is [1], for cache_positions, size is [1, max_seq_len]
+  metadata[llm::kUseCachePositions] = second_input_info.sizes().size() == 2;
```
return metadata?

mergennachin commented on examples/models/phi-3-mini/main.cpp:
```diff
@@ -42,9 +45,17 @@ int main(int32_t argc, char** argv) {
-
-  runner.generate(prompt, seq_len);
+  std::unique_ptr<tokenizers::Tokenizer> tokenizer =
+      std::make_unique<tokenizers::Llama2cTokenizer>();
+  tokenizer->load(tokenizer_path);
+  std::cout << "Tokenizer loaded, eos_id = " << tokenizer->eos_tok()
+            << std::endl;
+  auto runner = executorch::extension::llm::create_text_llm_runner(
+      model_path, std::move(tokenizer));
+
+  runner->generate(
+      prompt,
```
nice!!

mergennachin commented on examples/models/phi-3-mini/CMakeLists.txt:
```diff
@@ -39,16 +41,13 @@ endif()
 target_link_libraries(
-  phi_3_mini_runner PRIVATE executorch extension_module_static extension_tensor
-                            optimized_native_cpu_ops_lib xnnpack_backend gflags
+  phi_3_mini_runner PUBLIC optimized_native_cpu_ops_lib xnnpack_backend gflags extension_llm_runner
```
<3

shoumikhin commented on extension/tensor/tensor_ptr_maker.cpp:
```diff
@@ -186,5 +186,59 @@ TensorPtr randint_strided(
+  // Create a 1D tensor with the calculated size
+  auto tensor = empty_strided({numel}, {1}, type, dynamism);
+
+  // Fill the tensor with values from start to end with step
```
nit: I'd skip some of the comments if not all to not parrot the code, but maybe it's just me :)

shoumikhin commented on examples/models/phi-3-mini/main.cpp:
```diff
@@ -6,9 +6,12 @@
  * LICENSE file in the root directory of this source tree.
  */
 
+#include <executorch/extension/llm/runner/text_llm_runner.h>
```
nit: may like to keep headers sorted: std lib first, then 1st party

larryliu0820 replied:
This is properly formatted (not sure why this way though)

mergennachin commented:
@larryliu0820 do you plan to have this as part of 0.7?

larryliu0820 committed:
Phi3 runner uses TextLLMRunner

As titled, this PR started to use `TextLLMRunner` to run phi-3-mini.

Eager model comes from Huggingface, not using kv cache as custom op
because it is only being supported on Optimum-executorch repo.

Performance may not be the best.
